File: web/src/adr036.ts
// TypeScript port of the server's internal/adr036 package. It must produce
// byte-identical canonical message + ADR-036 sign-doc to the Go code, so that
// (a) wallet signatures verify server-side and (b) we can re-verify stored
// ratings in the browser. Pinned by __tests__/adr036.test.ts against Go output.

import { Secp256k1, Secp256k1Signature, sha256, ripemd160 } from "@cosmjs/crypto";
import { fromBase64, toBase64, toBech32, toUtf8 } from "@cosmjs/encoding";
import type { Feedback, FeedbackRecord, Role } from "./types";

// goJSONString replicates Go's encoding/json string escaping (HTMLEscape on):
// short escapes for " \ \n \r \t; <, >, & and U+2028/U+2029 as \uXXXX; other
// control chars < 0x20 as \u00XX; everything else verbatim (UTF-8).
function goJSONString(s: string): string {
  let out = '"';
  for (const ch of s) {
    const c = ch.codePointAt(0)!;
    switch (ch) {
      case '"': out += '\\"'; break;
      case "\\": out += "\\\\"; break;
      case "\n": out += "\\n"; break;
      case "\r": out += "\\r"; break;
      case "\t": out += "\\t"; break;
      case "<": out += "\\u003c"; break;
      case ">": out += "\\u003e"; break;
      case "&": out += "\\u0026"; break;
      default:
        if (c === 0x2028 || c === 0x2029 || c < 0x20) {
          out += "\\u" + c.toString(16).padStart(4, "0");
        } else {
          out += ch;
        }
    }
  }
  return out + '"';
}

// canonicalFeedbackBytes returns the deterministic JSON that is actually signed,
// with lexicographically sorted keys and no whitespace (matches Go's
// marshalCanonical: comment, deployer, dseq, gseq, issued_at, oseq, provider,
// role, score, v).
export function canonicalFeedbackBytes(f: Feedback): string {
  return (
    "{" +
    '"comment":' + goJSONString(f.comment) + "," +
    '"deployer":' + goJSONString(f.deployer) + "," +
    '"dseq":' + goJSONString(f.dseq) + "," +
    '"gseq":' + goJSONString(f.gseq) + "," +
    '"issued_at":' + goJSONString(f.issued_at) + "," +
    '"oseq":' + goJSONString(f.oseq) + "," +
    '"provider":' + goJSONString(f.provider) + "," +
    '"role":' + goJSONString(f.role) + "," +
    '"score":' + String(f.score) + "," +
    '"v":' + goJSONString(f.v) +
    "}"
  );
}

// buildSignDoc wraps the canonical message in the amino ADR-036 StdSignDoc, the
// exact bytes a wallet hashes and signs (and what the server reconstructs).
export function buildSignDoc(canonical: string, signer: string): string {
  const data = toBase64(toUtf8(canonical));
  return (
    '{"account_number":"0","chain_id":"","fee":{"amount":[],"gas":"0"},"memo":"",' +
    '"msgs":[{"type":"sign/MsgSignData","value":{"data":"' + data + '","signer":"' + signer + '"}}],' +
    '"sequence":"0"}'
  );
}

// akashAddressFromPubkey derives the bech32 akash address from a compressed
// secp256k1 public key: ripemd160(sha256(pubkey)) -> bech32("akash").
export function akashAddressFromPubkey(pubkey: Uint8Array): string {
  return toBech32("akash", ripemd160(sha256(pubkey)));
}

function feedbackFromRecord(r: FeedbackRecord): Feedback {
  return {
    v: "akash-reputation/v1",
    role: r.role as Role,
    provider: r.provider,
    deployer: r.deployer,
    dseq: r.dseq,
    gseq: r.gseq,
    oseq: r.oseq,
    score: r.score,
    comment: r.comment,
    issued_at: r.issued_at,
  };
}

// verifyRecord re-checks a stored rating client-side: the signature must be a
// valid secp256k1 signature by `pubkey` over the rating's sign-doc, and the
// address derived from `pubkey` must equal the rating's author.
export async function verifyRecord(r: FeedbackRecord): Promise<boolean> {
  try {
    const pubkey = fromBase64(r.pubkey);
    const sig = fromBase64(r.signature);
    if (sig.length !== 64) return false;
    if (akashAddressFromPubkey(pubkey) !== r.author) return false;
    const canonical = canonicalFeedbackBytes(feedbackFromRecord(r));
    const doc = buildSignDoc(canonical, r.author);
    const hash = sha256(toUtf8(doc));
    return await Secp256k1.verifySignature(Secp256k1Signature.fromFixedLength(sig), hash, pubkey);
  } catch {
    return false;
  }
}

File: internal/adr036/adr036.go
// Package adr036 implements Cosmos SDK ADR-036 off-chain arbitrary-message
// signing and verification for the Akash reputation service.
//
// A feedback message is canonicalised to deterministic JSON, wrapped in the
// well-known amino StdSignDoc with a single sign/MsgSignData message, hashed
// with SHA-256 and signed with secp256k1. This is byte-for-byte compatible with
// Keplr's signArbitrary, so browser wallets can produce signatures this package
// verifies, and vice versa.
//
// The same Sign/Verify pair is used by both the server (verify) and the
// repute-sign CLI (sign), so the two halves can never drift apart.
package adr036

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"sort"

	"github.com/cosmos/btcutil/bech32"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
	"golang.org/x/crypto/ripemd160" //nolint:staticcheck // cosmos address derivation requires ripemd160
)

// AddrHRP is the bech32 human-readable prefix for Akash account addresses.
const AddrHRP = "akash"

// MessageVersion is the canonical version tag embedded in every feedback message.
const MessageVersion = "akash-reputation/v1"

// Role identifies which party authored a feedback message.
type Role string

const (
	// RoleProvider means the provider operator is rating a deployer.
	RoleProvider Role = "provider"
	// RoleDeployer means a deployer is rating the provider.
	RoleDeployer Role = "deployer"
)

// Feedback is the canonical, signable reputation statement. The author is the
// party named by Role (provider rates deployer, or deployer rates provider) and
// must equal the address derived from the signing key.
type Feedback struct {
	Version  string `json:"v"`
	Role     Role   `json:"role"`
	Provider string `json:"provider"`
	Deployer string `json:"deployer"`
	DSeq     string `json:"dseq"`
	GSeq     string `json:"gseq"`
	OSeq     string `json:"oseq"`
	Score    int    `json:"score"` // +1 or -1
	Comment  string `json:"comment"`
	IssuedAt string `json:"issued_at"` // RFC3339 UTC
}

// Author returns the address that must have produced the signature for this
// feedback to be valid: the provider for provider-role, the deployer otherwise.
func (f Feedback) Author() string {
	if f.Role == RoleProvider {
		return f.Provider
	}
	return f.Deployer
}

// Subject returns the address the feedback is about (the counterparty).
func (f Feedback) Subject() string {
	if f.Role == RoleProvider {
		return f.Deployer
	}
	return f.Provider
}

// Validate checks structural invariants independent of any signature.
func (f Feedback) Validate() error {
	if f.Version != MessageVersion {
		return fmt.Errorf("unexpected version %q, want %q", f.Version, MessageVersion)
	}
	if f.Role != RoleProvider && f.Role != RoleDeployer {
		return fmt.Errorf("invalid role %q", f.Role)
	}
	if f.Score != 1 && f.Score != -1 {
		return fmt.Errorf("score must be +1 or -1, got %d", f.Score)
	}
	for name, addr := range map[string]string{"provider": f.Provider, "deployer": f.Deployer} {
		if _, err := DecodeAddress(addr); err != nil {
			return fmt.Errorf("invalid %s address: %w", name, err)
		}
	}
	if f.DSeq == "" || f.GSeq == "" || f.OSeq == "" {
		return fmt.Errorf("dseq/gseq/oseq must all be set")
	}
	if f.IssuedAt == "" {
		return fmt.Errorf("issued_at must be set")
	}
	return nil
}

// CanonicalBytes returns the deterministic JSON encoding of the feedback that is
// actually signed (as the ADR-036 message "data"). Keys are sorted, no
// whitespace, so independent implementations produce identical bytes.
func (f Feedback) CanonicalBytes() ([]byte, error) {
	m := map[string]any{
		"v":         f.Version,
		"role":      string(f.Role),
		"provider":  f.Provider,
		"deployer":  f.Deployer,
		"dseq":      f.DSeq,
		"gseq":      f.GSeq,
		"oseq":      f.OSeq,
		"score":     f.Score,
		"comment":   f.Comment,
		"issued_at": f.IssuedAt,
	}
	return marshalCanonical(m)
}

// SignDoc builds the ADR-036 amino StdSignDoc sign bytes for the given feedback
// and signer address. This is the exact byte string a wallet hashes and signs.
func SignDoc(f Feedback, signer string) ([]byte, error) {
	data, err := f.CanonicalBytes()
	if err != nil {
		return nil, err
	}
	doc := map[string]any{
		"account_number": "0",
		"chain_id":       "",
		"fee":            map[string]any{"amount": []any{}, "gas": "0"},
		"memo":           "",
		"msgs": []any{map[string]any{
			"type": "sign/MsgSignData",
			"value": map[string]any{
				"data":   base64.StdEncoding.EncodeToString(data),
				"signer": signer,
			},
		}},
		"sequence": "0",
	}
	return marshalCanonical(doc)
}

// Sign produces a 64-byte [R||S] secp256k1 signature over the ADR-036 sign doc
// for the feedback, using priv. The returned signer address is derived from the
// key and must match feedback.Author() for the signature to be accepted.
func Sign(f Feedback, priv *secp256k1.PrivateKey) (sig []byte, pubKey []byte, signer string, err error) {
	pubKey = priv.PubKey().SerializeCompressed()
	signer, err = AddressFromPubKey(pubKey)
	if err != nil {
		return nil, nil, "", err
	}
	doc, err := SignDoc(f, signer)
	if err != nil {
		return nil, nil, "", err
	}
	hash := sha256.Sum256(doc)
	// SignCompact returns [V || R || S]; cosmos uses the low-S 64-byte R||S.
	compact := ecdsa.SignCompact(priv, hash[:], true)
	return compact[1:], pubKey, signer, nil
}

// Verify checks that sig is a valid secp256k1 signature by pubKey over the
// ADR-036 sign doc for f, and that the address derived from pubKey equals
// f.Author(). It returns the verified author address on success.
func Verify(f Feedback, pubKey, sig []byte) (author string, err error) {
	if len(sig) != 64 {
		return "", fmt.Errorf("signature must be 64 bytes, got %d", len(sig))
	}
	signer, err := AddressFromPubKey(pubKey)
	if err != nil {
		return "", fmt.Errorf("derive signer address: %w", err)
	}
	if signer != f.Author() {
		return "", fmt.Errorf("signer %s is not the message author %s (role=%s)", signer, f.Author(), f.Role)
	}
	pk, err := secp256k1.ParsePubKey(pubKey)
	if err != nil {
		return "", fmt.Errorf("parse pubkey: %w", err)
	}
	doc, err := SignDoc(f, signer)
	if err != nil {
		return "", err
	}
	hash := sha256.Sum256(doc)
	var r, s secp256k1.ModNScalar
	if overflow := r.SetByteSlice(sig[:32]); overflow {
		return "", fmt.Errorf("signature R overflows curve order")
	}
	if overflow := s.SetByteSlice(sig[32:]); overflow {
		return "", fmt.Errorf("signature S overflows curve order")
	}
	if !ecdsa.NewSignature(&r, &s).Verify(hash[:], pk) {
		return "", fmt.Errorf("signature verification failed")
	}
	return signer, nil
}

// AddressFromPubKey derives the bech32 akash account address from a compressed
// secp256k1 public key, using the standard cosmos derivation
// (ripemd160(sha256(pubkey))).
func AddressFromPubKey(pubKey []byte) (string, error) {
	if _, err := secp256k1.ParsePubKey(pubKey); err != nil {
		return "", fmt.Errorf("invalid pubkey: %w", err)
	}
	sha := sha256.Sum256(pubKey)
	h := ripemd160.New()
	if _, err := h.Write(sha[:]); err != nil {
		return "", err
	}
	addr := h.Sum(nil) // 20 bytes
	conv, err := bech32.ConvertBits(addr, 8, 5, true)
	if err != nil {
		return "", err
	}
	return bech32.Encode(AddrHRP, conv)
}

// DecodeAddress validates an akash bech32 address and returns the 20-byte payload.
func DecodeAddress(addr string) ([]byte, error) {
	hrp, data, err := bech32.Decode(addr, 1023)
	if err != nil {
		return nil, err
	}
	if hrp != AddrHRP {
		return nil, fmt.Errorf("unexpected hrp %q, want %q", hrp, AddrHRP)
	}
	conv, err := bech32.ConvertBits(data, 5, 8, false)
	if err != nil {
		return nil, err
	}
	if len(conv) != 20 {
		return nil, fmt.Errorf("address payload is %d bytes, want 20", len(conv))
	}
	return conv, nil
}

// marshalCanonical marshals v to JSON with lexicographically sorted object keys
// and no insignificant whitespace, matching amino's signable JSON encoding.
func marshalCanonical(v any) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	var generic any
	if err := json.Unmarshal(raw, &generic); err != nil {
		return nil, err
	}
	return sortedMarshal(generic)
}

func sortedMarshal(v any) ([]byte, error) {
	switch t := v.(type) {
	case map[string]any:
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		buf := []byte{'{'}
		for i, k := range keys {
			if i > 0 {
				buf = append(buf, ',')
			}
			kb, err := json.Marshal(k)
			if err != nil {
				return nil, err
			}
			buf = append(buf, kb...)
			buf = append(buf, ':')
			vb, err := sortedMarshal(t[k])
			if err != nil {
				return nil, err
			}
			buf = append(buf, vb...)
		}
		return append(buf, '}'), nil
	case []any:
		buf := []byte{'['}
		for i, e := range t {
			if i > 0 {
				buf = append(buf, ',')
			}
			eb, err := sortedMarshal(e)
			if err != nil {
				return nil, err
			}
			buf = append(buf, eb...)
		}
		return append(buf, ']'), nil
	default:
		return json.Marshal(v)
	}
}
